src/zedbit/mailx/Mailx.java

```diff
@@ -0,0 +1,506 @@
+
+    } // getStaticLinks
+
+    /**
+    * visitDynamicLinks finds clickable items in the current page 
+    * and executes a simulated HtmlUnit click action on them. It then
+    * passes the new current page to  <code> crawl(HtmlPage page)</code>
+    */
+    private static void visitDynamicLinks (HtmlPage page, String pageUrl) {
+
+        // >>> Future upgrades! Here more logic could be added for buttons, RoR dynamic content, ASP, etc... *****/
+
+        // Now get list of all elements that Angular uses for clickable dynamic links
+        final List<?> clickables = page.getByXPath("//*[contains(@ng-click,'changeRoute')]");
+        printlnT("At: " + pageUrl);
+        printlnT("elements w/ dynamic click hooks: " + clickables.size());
+        clickables.forEach((i) -> { printlnT(i.toString());});
+        printlnT("-------------------------");
+
+        // Now here we click on the dynamic content, return via "Back" and keep collecting static content
+        clickables.forEach((i) -> {
+            printlnT("HtmlUnit browser getting page via simulated click of " + i.toString());
+            HtmlPage newPage;
+            try { newPage = ((DomElement) i).click(); }
+            catch (Exception e) {
+                System.err.println("Click action on " + i.toString() + " failed. Skipping it");
+                if (verbose) { e.printStackTrace(); }
+                return;
+            }
+            // Wait a bit for the page's JavaScript and HtmlUnit to catch up.
+            webClient.waitForBackgroundJavaScript(100);
+            crawl (newPage);
+    
+        });
+    } // End visitDynamicLinks
+
+    /**
+    * visitStaticLinks visits all static links we have accumulated after examining
+    * the current page 
+    */
+    private static void visitStaticLinks (HtmlPage page, String pageUrl) {
+        // Now we recurse on the static content that we accumulated
+        try {
+            urlsReachable.forEach((i) -> {
+                printlnT("---- traversing static link: " + i);
+                if (urlsVisited.contains(i) == false) {
+                    // We visit a URL we have not visited
+                    printlnT("---- will crawl " + i);
+                    webClient.waitForBackgroundJavaScript(100);
+                    crawl(i);
+                }
+            });
+            // Now we can empty the set, as we have visited them all, or tried to
+            urlsReachable.clear();
+        } catch (Exception e) {
+            printlnV("Cannot iterate further! Async JavaScript HtmlUnit problems?!");
+            urlsReachable.clear();
+            if (trace) { e.printStackTrace(); }
+            return;
+        }
+    } // end Visit Static Links
+
+
+
+    /**
+    * filterByType removes links to static links such as pngs, jpgs, icos, and
+    * under types that we cannot currently handle, which are stored in the regex
+    * skipP 
+    */
+    private static void filterByType(Set<String> urls) {
```

Comment:
Hi Chris, this is a really good point. I have rewritten filterByType() so that it does not do its job as a side effect, and now it efficiently builds the result in one pass vs. two. Now getStaticLinks() ends with this line

return filterByType(linksAsSet);

I also rewrote getStaticLinks so that it is DRY, as the logic for HREFs and ANCHORs was almost the same. Rather than duplicating, I use an enum. getStaticLinks() now returns a set, which made visitStaticLinks() much more straightforward and smaller.